File: C/347. Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        //Key: number, Value: count
        unordered_map<int, int> map;

        for(int i = 0; i < nums.size(); i++){
            map[nums[i]]++;
        }

        vector<int> result;

        for(int i = 0; i < k; i++){
            int max = 0;
            int maxKey;

            for(auto it = map.begin(); it != map.end(); it++){
                if(it->second > max){
                    max = it->second;
                    maxKey = it->first;
                }
            }
            map.erase(maxKey);
            result.push_back(maxKey);
        }

        return result;
    }
};

File: Java/Arrays & Hashing/347. Top K Frequent Elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        //Key: Number, Value: Count
        HashMap<Integer, Integer> count = new HashMap<>();

        //Add elements from nums to HashMap
        for(int i = 0; i < nums.length; i++){
            count.merge(nums[i], 1, Integer::sum);
        }

        //Convert array to ArrayList based on values, then reverse so descending order, then split based on k
        ArrayList<Integer> list = new ArrayList<>(count.values());
        Collections.sort(list, Collections.reverseOrder());
        ArrayList<Integer> list2 = new ArrayList<>(list.subList(0,k));

        int[] result = new int[list2.size()];
        int j = 0;

        //For each key/Value pair
        for(var entry : count.entrySet()){
            if(list2.contains(entry.getValue())){
                result[j++] = entry.getKey();
            }
        }

        return result;
    }
}
